File: Java/Gold/1504.java
import java.util.*;
import java.io.*;

public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static ArrayList<ArrayList<Vertex>> adj = new ArrayList<>();
    private static int[] checkPoint;
    private static int V;
    private static long[] dist;
    public static void main(String[] args) throws IOException {
        settingInput();
        System.out.println(visitVertexWithOrder(checkPoint[0], checkPoint[1]));
        br.close();
    }

    private static void settingInput() throws IOException {
        int[] VE = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        V = VE[0];
        for(int i=0;i<VE[0]+1;i++) adj.add(new ArrayList<>());
        for(int i=0;i<VE[1];i++) {
            int[] edge = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            adj.get(edge[0]).add(new Vertex(edge[1], edge[2]));
            adj.get(edge[1]).add(new Vertex(edge[0], edge[2]));
        }
        dist = new long[VE[0]+1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        checkPoint = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
    }

    private static void dijkstra(int startVertex) {
        PriorityQueue<Vertex> pq = new PriorityQueue<>(Comparator.comparingLong(a -> a.weight));
        dist[startVertex] = 0;
        pq.offer(new Vertex(startVertex, 0));

        while(!pq.isEmpty()) {
            Vertex currentVertex = pq.poll();

            if (dist[currentVertex.number] < currentVertex.weight) continue;
            adj.get(currentVertex.number).forEach( (next) -> {
                long nextDist = currentVertex.weight + next.weight;
                if (dist[next.number] > nextDist) {
                    dist[next.number] = nextDist;
                    pq.offer(new Vertex(next.number, nextDist));
                }
            });

        }
    }

    private static long visitVertexWithOrder(int firstVertex, int secondVertex) {
        long startToFirstVertex = 0;
        long startToSecondVertex = 0;
        long firstVertexToSecondVertex = 0;
        long firstVertexToEnd = 0;
        long secondVertexToEnd = 0;

        dijkstra(firstVertex);
        startToFirstVertex = dist[1];
        firstVertexToSecondVertex = dist[secondVertex];
        firstVertexToEnd = dist[V];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dijkstra(secondVertex);
        startToSecondVertex = dist[1];
        secondVertexToEnd = dist[V];

        long routeA = startToFirstVertex + firstVertexToSecondVertex + secondVertexToEnd;
        long routeB = startToSecondVertex + firstVertexToSecondVertex + firstVertexToEnd;

        if (routeA >= Integer.MAX_VALUE && routeB >= Integer.MAX_VALUE) return -1;
        else return Long.min(routeA, routeB);
    }

}

class Vertex {
    int number;
    long weight;

    Vertex(int number, long weight) {
        this.number = number;
        this.weight = weight;
    }
}

File: Archived2/Kotlin/Gold/1504.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.collections.ArrayList
import kotlin.math.min

lateinit var adj: ArrayList<ArrayList<Vertex>>
lateinit var dist: IntArray
val br = BufferedReader(InputStreamReader(System.`in`))
var V = 0
lateinit var checkPoint: List<Int>

fun main() {
    settingInput()
    println(visitVertexWithOrder())
    br.close()
}

fun settingInput() {
    val VE = br.readLine().split(" ").map { it.toInt() }
    V = VE[0]
    adj = ArrayList()
    repeat(V+1) { adj.add(ArrayList()) }
    repeat(VE[1]) {
        val edge = br.readLine().split(" ").map { it.toInt() }
        adj.get(edge[0]).add(Vertex(edge[1], edge[2]))
        adj.get(edge[1]).add(Vertex(edge[0], edge[2]))
    }
    dist = IntArray(V+1)
    dist.fill(Int.MAX_VALUE)
    checkPoint = br.readLine().split(" ").map { it.toInt() }
}

fun dijkstra(start: Int) {
    val pq = PriorityQueue { a: Vertex, b: Vertex -> a.weight - b.weight }
    dist[start] = 0
    pq.offer(Vertex(start, 0))

    while(pq.isNotEmpty()) {
        val currentVertex = pq.poll()

        if(dist[currentVertex.number] < currentVertex.weight) continue
        adj.get(currentVertex.number).forEach {
            val nextDist = currentVertex.weight + it.weight
            if (dist[it.number] > nextDist) {
                dist[it.number] = nextDist
                pq.offer(Vertex(it.number, nextDist))
            }
        }
    }
}

fun visitVertexWithOrder(): Long {

    var startToFirstVertex = 0L
    var startToSecondVertex = 0L
    var firstVertexToSecondVertex = 0L
    var firstVertexToEnd = 0L
    var secondVertexToEnd = 0L

    dijkstra(checkPoint[0])
    startToFirstVertex = dist[1].toLong()
    firstVertexToEnd = dist[V].toLong()
    firstVertexToSecondVertex = dist[checkPoint[1]].toLong()
    dist.fill(Int.MAX_VALUE)
    dijkstra(checkPoint[1])
    startToSecondVertex = dist[1].toLong()
    secondVertexToEnd = dist[V].toLong()

    val routeA: Long = startToFirstVertex + firstVertexToSecondVertex + secondVertexToEnd
    val routeB: Long = startToSecondVertex + firstVertexToSecondVertex + firstVertexToEnd

    return if (routeA >= Int.MAX_VALUE && routeB >= Int.MAX_VALUE) -1
    else min(routeA, routeB)
}

data class Vertex(val number: Int, val weight: Int)